File: sdk-go/volumes_test.go
package boxkite

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"testing"
)

func TestCreateVolume_SendsLabelAndSizeGB(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/volumes" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["label"] != "training-data" || body["size_gb"] != float64(20) {
			t.Errorf("unexpected body: %v", body)
		}
		writeJSON(t, w, 202, `{"id": "vol-1", "label": "training-data", "size_gb": 20, "status": "queued", "pvc_name": null, "failure_reason": null, "created_at": "2026-01-01T00:00:00Z"}`)
	})
	defer closeServer()

	volume, err := client.CreateVolume(context.Background(), CreateVolumeRequest{Label: Ptr("training-data"), SizeGB: 20})
	if err != nil {
		t.Fatalf("CreateVolume: %v", err)
	}
	if volume.ID != "vol-1" || volume.Status != "queued" {
		t.Errorf("unexpected volume: %+v", volume)
	}
}

func TestGetVolume(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/volumes/vol-1" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"id": "vol-1", "label": "training-data", "size_gb": 20, "status": "ready", "pvc_name": "boxkite-volume-vol-1", "failure_reason": null, "created_at": "2026-01-01T00:00:00Z"}`)
	})
	defer closeServer()

	volume, err := client.GetVolume(context.Background(), "vol-1")
	if err != nil {
		t.Fatalf("GetVolume: %v", err)
	}
	if volume.Status != "ready" || volume.PVCName == nil {
		t.Errorf("unexpected volume: %+v", volume)
	}
}

func TestListVolumes(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, 200, `[{"id": "vol-1", "label": null, "size_gb": 10, "status": "ready", "pvc_name": "pvc-1", "failure_reason": null, "created_at": "2026-01-01T00:00:00Z"}]`)
	})
	defer closeServer()

	volumes, err := client.ListVolumes(context.Background())
	if err != nil {
		t.Fatalf("ListVolumes: %v", err)
	}
	if len(volumes) != 1 || volumes[0].ID != "vol-1" {
		t.Errorf("unexpected volumes: %+v", volumes)
	}
}

func TestDeleteVolume(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete || r.URL.Path != "/v1/volumes/vol-1" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(204)
	})
	defer closeServer()

	if err := client.DeleteVolume(context.Background(), "vol-1"); err != nil {
		t.Fatalf("DeleteVolume: %v", err)
	}
}

File: sdk-rust/tests/volumes_test.rs
mod common;

use boxkite_client::CreateVolumeOptions;
use serde_json::json;
use wiremock::matchers::{body_json, method, path};
use wiremock::{Mock, ResponseTemplate};

#[tokio::test]
async fn create_volume_sends_size_and_label() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/volumes"))
        .and(body_json(json!({"size_gb": 50.0, "label": "shared-data"})))
        .respond_with(ResponseTemplate::new(202).set_body_json(json!({
            "id": "vol_1", "label": "shared-data", "status": "queued", "created_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let volume = client
        .create_volume(50.0, CreateVolumeOptions::new().label("shared-data"))
        .await
        .expect("create_volume should succeed");

    assert_eq!(volume.id, "vol_1");
    assert_eq!(volume.status, "queued");
}

#[tokio::test]
async fn create_volume_omits_label_when_unset() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/volumes"))
        .and(body_json(json!({"size_gb": 10.0})))
        .respond_with(ResponseTemplate::new(202).set_body_json(json!({
            "id": "vol_2", "status": "queued", "created_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    client
        .create_volume(10.0, CreateVolumeOptions::new())
        .await
        .expect("create_volume should succeed");
}

#[tokio::test]
async fn get_volume_parses_ready_status() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/volumes/vol_1"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "id": "vol_1", "label": "shared-data", "size_gb": 50.0, "status": "ready",
            "pvc_name": "boxkite-vol-1", "failure_reason": null, "created_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let volume = client
        .get_volume("vol_1")
        .await
        .expect("get_volume should succeed");
    assert_eq!(volume.status, "ready");
    assert_eq!(volume.pvc_name.as_deref(), Some("boxkite-vol-1"));
}

#[tokio::test]
async fn list_volumes_returns_populated_vec() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/volumes"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!([
            {"id": "vol_1", "size_gb": 10.0, "status": "ready", "created_at": "2026-01-01T00:00:00Z"}
        ])))
        .mount(&server)
        .await;

    let volumes = client
        .list_volumes()
        .await
        .expect("list_volumes should succeed");
    assert_eq!(volumes.len(), 1);
}

#[tokio::test]
async fn delete_volume_sends_delete() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("DELETE"))
        .and(path("/v1/volumes/vol_1"))
        .respond_with(ResponseTemplate::new(204))
        .mount(&server)
        .await;

    client
        .delete_volume("vol_1")
        .await
        .expect("delete_volume should succeed");
}
